packages/gatsby/src/utils/node-manifest.ts

```diff
@@ -60,21 +60,54 @@ const NODE_MANIFEST_FILE_LIMIT = getNodeManifestFileLimit()
  */
 async function findPageOwnedByNode({
   nodeId,
+  fullNode,
   slug,
 }: {
   nodeId: string
+  fullNode: IGatsbyNode
   slug?: string
 }): Promise<{
   page: INodeManifestPage
   foundPageBy: FoundPageBy
 }> {
   const state = store.getState()
   const { pages, nodes } = state
-  const { byNode } = state.queries
+  const { byNode, byConnection, trackedComponents } = state.queries
 
-  // the default page path is the first page found in
-  // node id to page query tracking
-  let pagePath = byNode?.get(nodeId)?.values()?.next()?.value
+  const { staticQueryComponents } = state
+  const nodeType = fullNode?.internal?.type
+
+  const firstPageWithNodeAsDataDependency =
+    // the first page found in node id to page query path tracking
+    byNode?.get(nodeId)?.values()?.next()?.value
+
+  const firstPagePathWithNodeInGraphQLListField =
+    // the first page that queries for a list of this node type.
+    // we don't currently store a list of node ids for connection fields to queries
+    // we just store the query id or page path mapped to the connected GraphQL typename.
+    byConnection?.get(nodeType)?.values()?.next()?.value
+
+  const defaultPagePath =
+    firstPageWithNodeAsDataDependency || firstPagePathWithNodeInGraphQLListField
+
+  let pagePath = defaultPagePath
```

Comment:
You're not doing anything with defaultPagePath so we can allocate only one.
```suggestion
  const firstPagePathWithNodeInGraphQLListField =
    // the first page that queries for a list of this node type.
    // we don't currently store a list of node ids for connection fields to queries
    // we just store the query id or page path mapped to the connected GraphQL typename.
    byConnection?.get(nodeType)?.values()?.next()?.value

  let pagePath =
    firstPageWithNodeAsDataDependency || firstPagePathWithNodeInGraphQLListField
```